first cut at screenshot config generator

theomichel opened the pull request:
Open issues / questions:

- do we need command line params? I figure they're unnecessary because this will basically be run a few times by one or two people. Note that it currently puts the output in ./config/.
- output is not as readable as the current YAMLs are, as I'm having trouble getting pyyaml to do nice formatting. These are the main issues:

1. newlines in overseerScript are not getting preserved through the read/write. 
1. no breaks between screenshots in the same file, which would be nice for separation
1. order of the dictionary items for a given screenshot (name, url, etc.)  is alphabetical, and probably not guaranteed. It would be nice to have "name" first, then url etc.
1. I'm using a list for the screenshots within a state, so each screenshot has a "-" before it. If we could fix all the above, this would be no problem imo.

There are other yaml load/dump libraries that we can try, or it would be simple to implement our own. I'll play around with other libraries, but lmk if you have ideas.

theomichel committed:
first cut at screenshot config generator

theomichel committed:
switched to manual output of yaml

Manual output allows finer control of the output. Could probably do this with a library, but why bother?

theomichel commented:
Second commit switches to manually crafting the YAML. This is super easy (although hacky) and allows full control over ordering and formatting (fixes #1-#3 above) without learning another library. Probably is the best way to go?

theomichel commented:
p.s. I realize that I'm doing terrible things with pascal casing in a python world. I'll switch that before merging. Just want to make sure I'm on the right track first.

theomichel commented on screenshots/generate_new_configs.py:
```diff
@@ -0,0 +1,94 @@
+        existing_config = yaml.safe_load(f)
+
+    # set up the output directory
+    destination_directory = "./config/" + team + "/"
```
probably should be config**s**

julia326 replied:
yeah agreed - I'd also use the Python os path stuff, so `destination_directory = os.path.join(os.path.dirname(__file__), 'configs', team)`

julia326 commented on screenshots/generate_new_configs.py:
```diff
@@ -0,0 +1,94 @@
+def main(args_list=None):
+
+    state_urls = load_urls_from_live_info('https://covidtracking.com/api/states/info.csv')
+    output_yamls("taco", state_urls, 'core_screenshot_config.yaml')
```
haha say more?

theomichel replied:
"Testing And Covid Outcomes" : )

theomichel replied:
I'm ok with switching to tao or tno, but will leave it as taco if no one objects.

julia326 replied:
tacos work!

julia326 commented on screenshots/generate_new_configs.py:
```diff
@@ -0,0 +1,94 @@
+        new_config_list = []
+        for urlname, urltext in urls.items():
+            
+            # if there's no URL, that means there was an empty cell in the spreadsheet
```
this is mostly a matter of personal preference to avoid indenting lots of code, but here I'd do the reverse check:
```
if not urltext:
  continue

existing_state_config = existing_config[urlname].get(state)
...
```

theomichel replied:
will do

julia326 commented on screenshots/generate_new_configs.py:
```diff
@@ -0,0 +1,94 @@
+    for state, urls in state_urls.items():
+
+        # open a file for the current state, creating the file if it doesn't exist
+        outfile = open(destination_directory + state + '.yaml', 'w+')
```
it's better to use a "with" here to avoid the need to close the file handler:

```
with open(os.path.join(destination_directory, state + '.yaml'), 'w'):
  new_config_list = []
  ...
```

theomichel replied:
Derp, will do

julia326 commented on screenshots/generate_new_configs.py:
```diff
@@ -0,0 +1,94 @@
+def load_urls_from_live_info(url):
+
+    content = requests.get(url).content
+    state_info_df = pd.read_csv(io.StringIO(content.decode('utf-8')), keep_default_na=False)
```
just curious why the `keep_default_na=False`? (otherwise this looks pretty similar to the way links are read in backup_to_s3.py which makes sense)

theomichel replied:
Pandas was converting the empty cell into NaN, which was just slightly weird to me... I thought empty cells should be empty strings, and this did that.

julia326 commented on screenshots/generate_new_configs.py:
```diff
@@ -0,0 +1,94 @@
+            if(urltext):
+                existing_state_config = existing_config[urlname].get(state)
+                
+                outfile.write("- name: " + urlname + "\n")
```
this is a small change, but I think we talked about having the state name in the config as well so that it's not just inferred from the filename. So the final YAML would look like this:
```
state: AK
links:
  - name: primary
    url: http://dhss.alaska.gov/dph/Epi/id/Pages/COVID-19/monitoring.aspx
    overseerScript: |
      page.manualWait();  await page.waitForDelay(30000);  page.done();
    message: stuff

  - name: secondary
    url: https://experience.arcgis.com/experience/ed1c874ca60b4c15ab09095a070065ca
...  
```

In terms of using pyyaml vs formatting things manually: From a larger standpoint, I'm all for easier, though having to indent things manually to format them correctly is a little eek-inducing. It seems like there is a way to wrangle pyyaml into formatting multi-line strings (e.g. https://stackoverflow.com/questions/8640959/how-can-i-control-what-scalar-form-pyyaml-uses-for-my-data) but I also totally understand if you don't want to rewrite all this. I can also help with figuring out how to get pyyaml to do what we want?

theomichel replied:
I don't remember that, but I'm cool with it. Otherwise we would infer it from the filename, but that can be problematic, I agree.

As for the formatting, I did some searching and experimenting and couldn't easily get the formatting that I wanted. I'll give the stackoverflow you sent a try, but if it still doesn't quite do it, I'd prefer to stick with manual since this is just a temporary script.

theomichel replied:
I wasn't able to get the control that I was looking for via that stackexchange link, or others that I found. (without switching libraries) I'm inclined to stay with the manual \ning because it allows the most control, and it's pretty simple.

julia326 reviewed (commented):
thanks Theo! Some initial questions/comments

theomichel committed:
updated with most PR feedback

The exception is that I'm going to try the yaml "presenter" stuff that julia suggested instead of print() hackery.

theomichel committed:
a few whitespace tweaks

Just making everything as human-friendly as possible. If the yaml parser has trouble with this, we can remove the line breaks.

theomichel commented:
@julia326 I've updated the PR with changes that I think address all the things. (although still with the \n formatting)

julia326 commented on screenshots/generate_new_configs.py:
```diff
@@ -0,0 +1,102 @@
+                outfile.write("- name: " + url_name + "\n")
+                outfile.write("  url: " + url_text + "\n")
+
+                if(existing_state_config): # if there is a config from the previous yaml, use that and add name and url
+                    if(existing_state_config.get("overseerScript")):
+                        overseerscript_string = existing_state_config["overseerScript"]
+                        if overseerscript_string[-1] == "\n": overseerscript_string = overseerscript_string[:-1] 
+                        overseerscript_string = overseerscript_string.replace("\n", "\n    ") # indent
+                        outfile.write("  overseerScript: |\n    " + overseerscript_string + "\n")
+                    if(existing_state_config.get("renderSettings")):
+                        outfile.write("  renderSettings: \n    " + yaml.dump(existing_state_config["renderSettings"], default_flow_style=False, indent=6))
```
I'm thinking it'd be useful to leave newlines between overseer script statements for readability, the way we have them now. It's annoying string munging, but this is the snippet I'd use (also adding in things like .strip() for removing whitespace, etc.):
```suggestion
                if existing_state_config:
                    if "overseerScript" in existing_state_config:
                        outfile.write("  overseerScript: |\n")
                        # strip newlines
                        overseer_script = existing_state_config["overseerScript"].strip()
                        # each statement on its own line
                        for command in overseer_script.split(';'):
                            command = command.strip()
                            if command:
                                outfile.write("    %s;\n" % command)
                    if "renderSettings" in existing_state_config:
                        outfile.write("  renderSettings: \n    " + yaml.dump(
                            existing_state_config["renderSettings"], default_flow_style=False, indent=6))
```

julia326 reviewed (commented):
Thanks @theomichel ! One thing about saving overseerScript contents (see comments), otherwise LGTM.

theomichel committed:
Update screenshots/generate_new_configs.py

Co-authored-by: Julia Kodysh <julia326@gmail.com>

theomichel commented:
@julia326 I committed your suggestion and tried it out locally. I generally like it, but it does have a weird side effect of leaving a dangling }); in cases where we're using page.evaluate() with an anonymous function passed in. Do you want to leave that as is, or also insert newlines after { to make it look better?

julia326 commented on screenshots/generate_new_configs.py:
```diff
@@ -0,0 +1,107 @@
+                        overseer_script = existing_state_config["overseerScript"].strip()
+                        # each statement on its own line
+                        for command in overseer_script.split(';'):
+                            command = command.strip()
```
Agreed on the issue with the dangling `});` string - this would fix it I think: 

```suggestion
                            command = command.strip()
                            if '{' in command:
                                command = command.replace('{', '{\n     ')
```

theomichel committed:
Update screenshots/generate_new_configs.py

Co-authored-by: Julia Kodysh <julia326@gmail.com>

julia326 reviewed (approved):
yay thank you @theomichel !
